### PR TITLE
by yanniboi: Move edit link out of block label callback.

### PR DESCRIPTION
**Pull request by yanniboi**
… label.

### CODE REVIEW

**Comment by andrewbelcher** [comment]
Hmm... I feel like there may be a better way to solve this than bake more logic into label. Perhaps the link generation needs to be removed from `::label` entirely?

**Comment by yanniboi** [comment]
I would be in favour of that. When dealing with blocks in the dnd context, there are dozens of times that label is called without the edit link, and only one use case when the edit link is desirable.

**Comment by yanniboi** [comment]
Any ideas of where to put it?

**Comment by andrewbelcher** [comment]
The edit link (and whether to show it in the title context) probably needs to be pulled into something separate which is then added in as part of the rendering, rather than adding in as part of the label proper. This may be fine straight as part of building the render array, or it could be added as a pre-process or similar. The form can then use `::label()` freely.

**Comment by andrewbelcher** [comment]
Definitely the right direction. I wonder if we want to make handle it with a separate property on the render array and handle the actual rendering later? We could then do something like have an interface for it and make it not specific to a block?

**Review by andrewbelcher — COMMENTED** [review]
OK I was thinking that the interface would give us the url or something and then we'd add that as a property to out block and use a pre render to actually put it in place. That was things altering the block can use title as it currently is without the confusion of edit links etc?

File: contacts.module [hunk 1]
```diff
@@ -160,6 +161,30 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
...
+ *
+ * @param $element
+ *   The block render array.
+ *
+ * @return array
+ *   The block render array.
+ */
+function contacts_dashboard_block_edit_link_pre_render($element) {
+  if (!empty($link = $element['#dashboard_label_edit_link'])) {
+    if ($element['content']['#title']) {
```
Comment (andrewbelcher) [inline contacts.module]: Excellent, though the `empty($bar = call())` looks very odd to me... Wonder if some extra brevity there would make it read nicer!

File: contacts.module [hunk 2]
```diff
@@ -160,6 +161,30 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
...
+      $element['content']['#title'] = new FormattableMarkup('@label [@link]', [
+        '@label' => $element['content']['#title'],
+        '@link' => $link->toString(),
+      ]);
+    }
+    else {
+      $element['content']['#title'] = $link->toString();
```
Comment (andrewbelcher) [inline contacts.module:183]: Hmm... What does it do if we have no title or aren't displaying it? Have we thought about the formatting etc? Also raises the question of whether the title is actually the appropriate place for this, or if we're better thinking of a specific UI?

File: src/Element/ContactTabContent.php [hunk 3]
```diff
@@ -94,6 +95,12 @@ public static function preRenderTabContent(array $element) {
           'content' => $block->build(),
         ];
 
+        // Add edit link to title.
+        if (in_array(DashboardBlockInterface::class, class_implements($block))) {
```
Comment (andrewbelcher) [inline src/Element/ContactTabContent.php]: Would `$block implements Interface` be better?

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
This is starting to look awesome, well done Yan!

File: contacts.module (hunk 2 above)
Comment (yanniboi) [reply contacts.module:183]: Ok, there are several config options to configure:

label_display: ('visible' or '0') - if set to 0, not title will be shown and no edit link shown
edit_link: ('title' or 'content') - if we are not showing a title, this should be set to content.
label: if this is empty but label display is set to visible then we get a default title from $block->build().

So I think we actually have all bases covered. We would have to explicitly set the '#title' attribute to NULL for the link to be rendered on its own, which is either on purpose (to only show the edit link) or hiding it wrong.

File: contacts.module (hunk 1 above)
Comment (yanniboi) [reply contacts.module]: Not sure what you mean, something like this?

```
  if (!empty($element['#dashboard_label_edit_link'])) {
    $link = $element['#dashboard_label_edit_link'];
    if ($element['content']['#title']) {
```

File: src/Element/ContactTabContent.php (hunk 3 above)
Comment (yanniboi) [reply src/Element/ContactTabContent.php]: I'm pretty sure what you put isn't valid php :p 

I'm following core convention with my syntax. See:

Drupal\Core\Annotation\ContextDefinition::__construct()
Drupal\Core\Entity\TypedData\EntityDataDefinition::getPropertyDefinitions()

Comment (yanniboi) [reply src/Element/ContactTabContent.php]: (except that they tend to write out the full namespace of the class as a string, which feels like overkill...)

Comment (andrewbelcher) [reply src/Element/ContactTabContent.php]: Sorry, I meant `instanceof`.

File: contacts.module (hunk 2 above)
Comment (andrewbelcher) [reply contacts.module:183]: Ok - I'm happy for now. Let's make sure we have a general UI review of the edit links at some point.

File: src/Plugin/DashboardBlockInterface.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Drupal\contacts\Plugin;
+
+/**
+ * Defines an interface for blocks placed on the Contacts Dashboard.
+ */
+interface DashboardBlockInterface {
+
+  const EDIT_LINK_TITLE = 'title';
+  const EDIT_LINK_CONTENT = 'content';
```
Comment (andrewbelcher) [inline src/Plugin/DashboardBlockInterface.php]: Let's comment these while we're at it :)

File: contacts.module (hunk 1 above)
Comment (andrewbelcher) [reply contacts.module]: Yeah, I think so. Or we can use a null coalesce or adding default properties, so:
```php
if ($title = $element['#dashboard_label_edit_link'] ?? FALSE)
```
or
```php
$element += ['#dashboard_label_edit_link' => FALSE]
if ($element['#dashboard_label_edit_link'])
```

File: src/Plugin/DashboardBlockInterface.php (hunk 4 above)
Comment (yanniboi) [reply src/Plugin/DashboardBlockInterface.php]: The NULL coalescing operator (??) was added in php7 (http://php.net/manual/en/migration70.new-features.php) so we should avoid using it in contrib if possible. Not to mention our dev server is 5.6 so breaks my environment :P

Comment (andrewbelcher) [reply src/Plugin/DashboardBlockInterface.php]: https://www.drupal.org/project/drupal/issues/2842431 they are already planning on dropping 5.6 support. I'd be perfectly happy requiring php 7... Let's see what Jamie thinks...

**Review by andrewbelcher — APPROVED** [review]

**Pull request merged**